src/libcore/iter.rs

```diff
@@ -1791,6 +1797,16 @@ impl<T, I> Iterator for Peekable<T, I> where I: Iterator<Item=T> {
 }
 
 #[stable]
+impl<T, I> ExactSizeIterator for Peekable<T, I> where I: ExactSizeIterator<Item = T> {
+    #[inline]
+    fn len(&self) -> usize {
+        // This is guarenteed to not overflow because `len()` must have been able to return a valid
+        // value before we peeked.
+        self.iter.len() + if self.peeked.is_some() { 1 } else { 0 }
+    }
+}
```

Comment:
Sorry, that was supposed to be `(uint::MAX, Some(uint::MAX))` (copy and paste error).

My basic point is that that if the number of items in an `ExactSizeIterator` once fit in a `uint` (required by the spec), it must always fit in a `uint` regardless of where those items are (moving one from `self.iter` to `self.peeked` doesn't change the total number of items).